packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,785 @@
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  log: MongoLogWriter;
+  pendingLogEvents: CallableFunction[] = [];
+
+  telemetryAnonymousId: string | undefined;
+  userId: string | undefined;
+  trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  isSetup = false;
+  hasMongoLogWriterInitialized = false;
+
+  constructor(
+    public bus: MongoshBus,
+    public analytics: MongoshAnalytics,
+    public providedTraits: { platform: string } & {
+      [key: string]: unknown;
+    },
+    mongoshVersion: string
+  ) {
+    const dummySink = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    this.log = new MongoLogWriter('', null, dummySink);
+
+    this.userTraits = {
+      ...providedTraits,
+      session_id: this.log.logId,
+    };
+
+    this.trackingProperties = {
+      mongosh_version: mongoshVersion,
+      session_id: this.log.logId,
+    };
+  }
+
+  public setup() {
+    this._setupBusEventListeners();
+    this.isSetup = true;
+  }
+
+  public setupLogger(logger: MongoLogWriter) {
+    if (!this.isSetup) {
+      throw new Error('Run setup() before setting up the log writer.');
+    }
+    this.log = logger;
+
+    this.userTraits.session_id = this.log.logId;
+    this.trackingProperties.session_id = this.log.logId;
+  }
+
+  public getTelemetryUserIdentity = (): MongoshAnalyticsIdentity => ({
+    anonymousId: this.telemetryAnonymousId ?? (this.userId as string),
+  });
+
+  /** Start processing pending events after mongosh:logger-initialized */
+  private onLoggerInitialized() {
+    this.hasMongoLogWriterInitialized = true;
+    for (const pendingEvent of this.pendingLogEvents) {
+      pendingEvent();
+    }
+    this.pendingLogEvents = [];
+
+    hookLogger(this.bus, this.log, 'mongosh', (uri) =>
```

Comment:
Worth noting that the `devtools-connect` hooks right now only get initialized after the actual logger has been initialized. the reason for that is that, at least with our current setup, we end up passing the dummy log into this function, causing a bunch of event listeners to listen to even with the old log even after we swap it (since the reference of `this.log` doesn't change in this external case). 
If we want to be able to do the pending events in these cases, we may need to consider i.e. updating `devtools-connect` to pass a function to retrieve to get the logger instead.

Reply:
> at least with our current setup, we end up passing the dummy log into this function, causing a bunch of event listeners to listen to even with the old log even after we swap it (since the reference of this.log doesn't change in this external case).

This is a hypothetical case, right? As if, this is what would happen if we were to call `hookLogger` from the beginning, correct?

I think that's fine then. If we do need to do something to address this, we could also pass an object to `hookLogger` that buffers up log events until the real log destination is available, then forward events to that destination, similar to how we handle the events here.

Reply:
Ah yeah I meant 

> with our current setup ***if we were to initialize them like we do all the other onBus listeners setup()**..